src/ripple/app/tx/impl/CancelTicket.cpp

```diff
@@ -41,49 +42,23 @@ class CancelTicket
 
     }
 
-    TER doApply () override
+    TER preCheck () override
     {
-        assert (mTxnAccount);
-
-        uint256 const ticketId = mTxn.getFieldH256 (sfTicketID);
-
-        SLE::pointer sleTicket = mEngine->view ().entryCache (ltTICKET, ticketId);
-
-        if (!sleTicket)
-            return tecNO_ENTRY;
-
-        auto const ticket_owner =
-            sleTicket->getFieldAccount160 (sfAccount);
-
-        bool authorized =
-            mTxnAccountID == ticket_owner;
-
-        // The target can also always remove a ticket
-        if (!authorized && sleTicket->isFieldPresent (sfTarget))
-            authorized = (mTxnAccountID == sleTicket->getFieldAccount160 (sfTarget));
+        // A Ticket may not be used to cancel a Ticket.  So a Sequence of
+        // zero (which indicates that a Ticket is being used) is malformed.
+        if (mTxn.getSequence () == 0)
+            return tefPAST_SEQ;
 
-        // And finally, anyone can remove an expired ticket
-        if (!authorized && sleTicket->isFieldPresent (sfExpiration))
-        {
-            std::uint32_t const expiration = sleTicket->getFieldU32 (sfExpiration);
-
-            if (mEngine->getLedger ()->getParentCloseTimeNC () >= expiration)
-                authorized = true;
-        }
-
-        if (!authorized)
-            return tecNO_PERMISSION;
-
-        std::uint64_t const hint (sleTicket->getFieldU64 (sfOwnerNode));
-
-        TER const result = mEngine->view ().dirDelete (false, hint,
-            getOwnerDirIndex (ticket_owner), ticketId, false, (hint == 0));
+        return Transactor::preCheck ();
+    }
 
-        adjustOwnerCount(mEngine->view(), mEngine->view().entryCache(
-            ltACCOUNT_ROOT, getAccountRootIndex(ticket_owner)), -1);
-        mEngine->view ().entryDelete (sleTicket);
+    TER doApply () override
+    {
+        assert (mTxnAccount);
 
-        return result;
+        // See if we can legitimately use the Ticket.  Note that canceling
+        // an expired Ticket is a tesSUCCESS.
+        return checkAndConsumeCancelTicket (mTxn, mTxnAccountID, mEngine);
```

Comment:
`checkAndConsumeCancelTicket` is called only from here and nowhere else, so why not just inline the body in here? What problem is the indirection solving?